Fix NCCL version parsing for 2.10.3

KZFB opened the pull request:
Fixes #{63445}

see https://github.com/pytorch/pytorch/issues/63445 for context

Goal: update getNcclVersion in c10d/NCCLUtils.cpp to match https://github.com/NVIDIA/nccl/blob/7e515921295adaab72adf56ea71a0fafb0ecb5f3/src/nccl.h.in#L22



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Commit:
<Replace this line with a title. Use 1 line only, 67 chars or
less>

Summary:
Fix NCCL version parsing for 2.10.3 (bootcamp T98941867)

Test Plan:
python3 test/distributed/test_c10d_nccl.py

Reviewers:
Shen Li

Subscribers:

Tasks:

Tags:

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

mrshenli reviewed (commented):
Looks good. Thanks for fixing!

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

Commit:
<Replace this line with a title. Use 1 line only, 67 chars or less>

Summary:
Fix NCCL version parsing for 2.10.3

Test Plan:
python3 test/distributed/test_c10d_nccl.py

Reviewers:
Shen Li

Subscribers:

Tasks:

Tags:

mrshenli reviewed (approved):
LGTM. Please import this PR and land from phab.

mrshenli reviewed (changes requested):
Test failures are real:

```
Aug 26 19:36:01 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/NCCLUtils.cpp:21:7: error: multi-line comment [-Werror=comment]
Aug 26 19:36:01        // logic reference: https://github.com/NVIDIA/nccl/blob\
Aug 26 19:36:01        ^
Aug 26 19:36:01 cc1plus: all warnings being treated as errors
Aug 26 19:36:01 caffe2/CMakeFiles/torch_cuda.dir/build.make:4571: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/NCCLUtils.cpp.o' failed
Aug 26 19:36:01 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/NCCLUtils.cpp.o] Error 1
Aug 26 19:36:01 make[2]: *** Waiting for unfinished jobs....
Aug 26 19:36:02 make[1]: *** [caffe2/CMakeFiles/torch_cuda.dir/all] Error 2
Aug 26 19:36:02 CMakeFiles/Makefile2:10166: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/all' failed
Aug 26 19:36:02 make: *** [all] Error 2
Aug 26 19:36:02 Makefile:159: recipe for target 'all' failed
Aug 26 19:36:02 Building wheel torch-1.10.0a0+git0551588
Aug 26 19:36:02 -- Building version 1.10.0a0+git0551588
```

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

Commit:
Fix NCCL version parsing for 2.10.3 #64028

Summary:
Fix NCCL version parsing for 2.10.3 #64028

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

mrshenli reviewed (approved):
Stamp to unblock. Please wait for tests.

mrshenli reviewed (dismissed):
Build still fails. I think you can leave the url in one line and tell the linter to ignore that line

```
Aug 26 21:30:21 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/NCCLUtils.cpp:21:7: error: multi-line comment [-Werror=comment]
Aug 26 21:30:21        // logic reference: https://github.com/NVIDIA/nccl/blob\
Aug 26 21:30:21        ^
Aug 26 21:30:21 cc1plus: all warnings being treated as errors
Aug 26 21:30:21 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/NCCLUtils.cpp.o] Error 1
Aug 26 21:30:21 make[2]: *** Waiting for unfinished jobs....
Aug 26 21:30:21 caffe2/CMakeFiles/torch_cuda.dir/build.make:4571: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/distributed/c10d/NCCLUtils.cpp.o' failed
Aug 26 21:30:21 make[1]: *** [caffe2/CMakeFiles/torch_cuda.dir/all] Error 2
Aug 26 21:30:21 CMakeFiles/Makefile2:10166: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/all' failed
Aug 26 21:30:21 make: *** [all] Error 2
Aug 26 21:30:21 Makefile:159: recipe for target 'all' failed
```

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

Commit:
Fix NCCL version parsing for 2.10.3 #64028

Summary:
replace url in multiline comment with a shorter url

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

Commit:
Consolidating NCCL version parsing functions #64060

Summary:
Consolidating NCCL version parsing functions #64060

Test Plan:
build and test with test/distributed/test_c10d_nccl.py

Reviewers:
Shen Li

Subscribers:

Tasks:

Tags:

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

Commit:
Consolidating NCCL version parsing functions #64060

Summary:
Consolidating NCCL version parsing functions #64060

Test Plan:
build and test with test/distributed/test_c10d_nccl.py

Reviewers:

Subscribers:

Tasks:

Tags:

mrshenli commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -11,16 +11,15 @@ std::string getNcclVersion() {
-    // can't compute the version if call did not return successfully or version
-    // code < 100 (corresponding to 0.1.0)
-    if (status != ncclSuccess || version < 100) {
+    int version = torch::cuda::nccl::version();
```
do you need to include the following header before using this API? 

```c++
#include <torch/csrc/cuda/nccl.h>
```

The header does not seem to include that file:

https://github.com/pytorch/pytorch/blob/79203f6cab65ce908e54d183cbba65b8d44bf727/torch/csrc/distributed/c10d/NCCLUtils.hpp#L5-L12

IIUC, the `nccl.h` is the 3rd-party nccl not `cuda/nccl`.

KZFB replied:
Added the header.

Did the following to get a clean build:
$ python setup.py clean
$ rm -rf build
$ python setup.py install

Then test:
$ python3 test/distributed/test_c10d_nccl.py       
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss
----------------------------------------------------------------------
Ran 90 tests in 0.002s

OK (skipped=90)

mrshenli reviewed (commented):
I'd vote for separating bug fix (in `NCCLUtils.cpp`) with better-engineering code consolidation (`cuda/nccl`). But I won't block this PR due to not separating those.

+1 to @pbelevich comments on testing. Pavel, any suggestion on how we should test this? Exposing `NCCL_MAJOR` to Python? Another possibility is to create c++ tests? E.g., by adding a test method to this file: https://github.com/pytorch/pytorch/blob/master/test/cpp/c10d/ProcessGroupNCCLTest.cpp

Commit:
Consolidating NCCL version parsing functions #64060

Summary:
Consolidating NCCL version parsing functions #64060

Test Plan:
python3 setup.py clean, then install to build, then test with test/distributed/test_c10d_nccl.py

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Merge branch 'master' of github.com:pytorch/pytorch

pbelevich commented:
>  Pavel, any suggestion on how we should test this?

I'd just extract parsing version from `std::string getNcclVersion()` to new method `std::string parseNcclVersion(int)` and would cover it with some tests

mrshenli commented:
> I'd just extract parsing version from std::string getNcclVersion() to new method std::string parseNcclVersion(int) and would cover it with some tests

Thanks @pbelevich. Any suggestion on where we should put the new tests? A new C++ test file or some existing C++ test file (e.g., `ProcessGroupNCCLTest.cpp`) or expose that API to Python and test there?

eqy commented:
Does this also require changing the Python side (as that was also changed in #62916)? https://github.com/pytorch/pytorch/blob/master/torch/cuda/nccl.py

eqy commented:
One reason I avoided using the upstream NCCL formatting internally was that it wasn't clear how long the upstream version format would stay the same, and the compile-time macros seemed to be more reliable: 
https://github.com/NVIDIA/nccl/issues/543
